Serialization refactor: macros for value/row serialization

piodul opened the pull request:
Introduce derive macros for both of the new serialization traits: `SerializeCql` and `SerializeRow`. They allow to serialize given Rust struct either into a UDT value or a list of values for a query.

The most significant difference from the previous `IntoUserType` and `ValueList` is that the new macros match the Rust struct fields to UDT fields or columns by name. This makes the macros less error-prone to use, especially in the UDT case.

For users which are already used to the old macros and are willing to guarantee the field order but do not want to pay performance cost associated with sorting the fields, a separate "flavor" is introduced. With that flavor enabled, the generated code expects that the field order is correct. The field names are still checked - if the expectation about the matching fields is broken serialization will just fail.

Depends on https://github.com/scylladb/scylla-rust-driver/pull/862

Refs: #802

## Pre-review checklist

<!--
    Make sure you took care of the issues on the list.
    Put 'x' into those boxes which apply.
    You can also create the PR now and click on all relevant checkboxes.
    See CONTRIBUTING.md for more details.
-->

- [x] I have split my patch into logically separate commits.
- [x] All commit messages clearly explain what they change and why.
- [x] I added relevant tests for new features and bug fixes.
- [x] All commits compile, pass static checks and pass test.
- [x] PR description sums up the changes and reasons why they should be introduced.
- [x] I have provided docstrings for the public items that I want to introduce.
- [ ] I have adjusted the documentation in `./docs/source/`.
- [x] I added appropriate `Fixes:` annotations to PR description.

piodul commented:
v2:
- Rebased on top of https://github.com/scylladb/scylla-rust-driver/pull/862 (the current PR now depends on it),
- Implemented a variant of the macros which expects the UDT fields/columns to be in a particular order, saving some CPU cycles by not having to look up fields by name.

piodul commented:
v2.1: address clippy's complaint

piodul commented:
v2.2:
- Fixed links in docstrings
- Updated the `Cargo.lock.msrv`

piodul commented:
> I left a few questions - they are about commits belonging to #862 which I (mistakenly) thought I was reviewing, sorry. I'll review 4 commits of this PR shortly

No problem, my bad - the description of the other PR was misleading.

Lorak-mmk commented on scylla-macros/src/serialize/cql.rs:
```diff
@@ -0,0 +1,281 @@
+                            }
+                        }
+                    )*
+                    other => return ::std::result::Result::Err(mk_typck_err(
+                        #crate_path::UdtTypeCheckErrorKind::UnexpectedFieldInDestination {
+                            field_name: <_ as ::std::clone::Clone>::clone(field_name),
+                        }
+                    )),
```
You don't seem to use `other` variable, isn't `_` enough? Or maybe you meant to use it?

piodul replied:
No, I'm not using it. I can change it to `_`.

I don't have a documentation link to support my thesis, but it looks like some warnings are not issued about the code generated by macros (e.g. about unused bindings).

Lorak-mmk commented on scylla-cql/src/macros.rs:
````diff
@@ -75,6 +75,68 @@ pub use scylla_macros::ValueList;
+/// or [`BuiltinSerializationError`](crate::types::serialize::row::BuiltinSerializationError)
+/// will be returned.
+///
+/// # Example
+///
+/// A UDT defined like this:
+///
+/// ```notrust
+/// CREATE TYPE ks.my_udt (a int, b text, c blob);
+/// ```
+///
+/// ...can be serialized using the following struct:
+///
+/// ```rust
+/// # use scylla_cql::macros::SerializeRow;
+/// #[derive(SerializeRow)]
+/// # #[scylla(crate = scylla_cql)]
+/// struct MyUdt {
+///     a: i32,
+///     b: Option<String>,
+///     c: Vec<u8>,
+/// }
+/// ```
````
This is probably not the example you meant to put here, it should be one of statement with bind markers / columns and corresponding struct, no?

piodul replied:
Yes. It was copied from the SerializeCql example but I didn't update the description and the CQL part. I'll fix.

Lorak-mmk commented on scylla-macros/src/serialize/row.rs:
```diff
@@ -0,0 +1,247 @@
+                            }
+                        },
+                    )*
+                    other => return ::std::result::Result::Err(mk_typck_err(
+                        #crate_path::BuiltinRowTypeCheckErrorKind::MissingValueForColumn {
+                            name: <_ as ::std::clone::Clone>::clone(&&spec.name),
+                        }
```
ditto

piodul commented on scylla-cql/src/types/serialize/value.rs:
```diff
@@ -283,30 +283,178 @@ impl<T: Value + ?Sized> SerializeCql for &T {
+
+    let mut builder = writer.into_value_builder();
+
+    let element_count: i32 = len.try_into().expect("too many values to fit in i32");
```
Hmm...

The interface I introduced in https://github.com/scylladb/scylla-rust-driver/pull/855 panics when we try to produce a value which overflows `i32`. Panicking here would be consistent with that (each item has at least 4 bytes, so the whole thing overflows). I know that the old interface does not do this, but my rationale back then was that

- `Vec::push` panics on overflow,
- In one of the iterations of the design, I was updating serialized length in destructors, and that could overflow - the only sensible way to handle that was to panic.

However:

- `Vec` uses `isize` and it's not possible currently to overflow it without running out of memory; it's not true for `i32`,
- The current design doesn't have custom destructor logic and `finish()` could be used to report the overflow error.

What do you think? Personally, I'm changing my mind about panicking and I think I should go over and remove the panic-on-overflow logic.

Lorak-mmk replied:
I think that the 2 occurences of `expect` in this files can be removed (because they happen in functions returning `SerializationError`), so they should be removed - because as you said previous interface didn't panic, but also just to avoid panics in a library code.

piodul replied:
I'll remove the two `expect`s, but I'll de-panic the row writers / cell writers / cell value builders. Some of their methods will start returning `Result` so there will be a large number of changes to this PR, but they will mostly be mechanical.

piodul replied:
This is now fixed in the other PR.

piodul commented on scylla-cql/src/types/serialize/value.rs:
```diff
@@ -512,7 +496,224 @@ impl<'a, T: SerializeCql + 'a> SerializeCql for &'a [T] {
+    }
+}
+
+fn fix_cql_value_name_in_err(err: SerializationError) -> SerializationError {
+    // Unfortunately, it seems that there is no `downcast` for Arc<dyn Error>.
+    // We need to clone the error if the name is not correct
+
+    let rust_name = std::any::type_name::<CqlValue>();
+
+    if let Some(err) = err.downcast_ref::<BuiltinTypeCheckError>() {
+        if err.rust_name != rust_name {
+            return Arc::new(BuiltinTypeCheckError {
+                rust_name,
+                ..err.clone()
+            });
+        }
+    }
+
+    if let Some(err) = err.downcast_ref::<BuiltinSerializationError>() {
+        if err.rust_name != rust_name {
+            return Arc::new(BuiltinSerializationError {
+                rust_name,
+                ..err.clone()
+            });
+        }
+    }
+
+    err
+}
```
I did it, but if `Arc::downcast` existed it would be possible to write it in a bit more clean way.

piodul commented on scylla-cql/src/macros.rs:
```diff
@@ -13,6 +13,158 @@ pub use scylla_macros::IntoUserType;
+/// # use scylla_cql::macros::SerializeRow;
+/// #[derive(SerializeRow)]
+/// # #[scylla(crate = scylla_cql)]
+/// struct MyUdt {
```
Renamed it to `MyValues`

piodul commented:
v4:
- rebased
- updated docstrings, yet again
- fixed the bug that caused macros not to work for structs with generics

@Lorak-mmk - I manually checked that the test you commented out in your PR now compiles, so you can rebase.

Lorak-mmk reviewed (approved):
Looks good - but maybe add a test for generics so that we have an actual test for that beside the code in `examples/`?

piodul commented:
> Looks good - but maybe add a test for generics so that we have an actual test for that beside the code in `examples/`?

Done.

v4.1: added some minimal tests for UDTs/rows with generics in them, mostly to see whether they compile or not
